### PR TITLE
Added support for AGE and REGEXP_MATCHES function and unparsing logic…

### PR DESCRIPTION
**Pull request by Kowsikraja**
… of TRUNCATE Function in PostgresqlSqlDialect

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -2204,10 +2219,10 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
 
   /** The "DATE_TRUNC(date, timeUnit)" function (BigQuery);
    * truncates a DATE value to the beginning of a timeUnit. */
-  @LibraryOperator(libraries = {BIG_QUERY, SPARK})
+  @LibraryOperator(libraries = {BIG_QUERY, SPARK, POSTGRESQL})
   public static final SqlFunction DATE_TRUNC =
       SqlBasicFunction.create("DATE_TRUNC",
-          ReturnTypes.ARG0_NULLABLE,
+          ReturnTypes.ARG0_EXCEPT_STRING_NULLABLE,
```
Comment (divyanshuS27) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2225]: we have changed this type for BigQuery and spark dialects too. Did we verify whether this is as expected for those dialects too?

File: core/src/main/java/org/apache/calcite/sql/dialect/PostgresqlSqlDialect.java [hunk 2]
```diff
@@ -113,6 +113,17 @@ public PostgresqlSqlDialect(Context context) {
         SqlParserPos.ZERO);
   }
 
+  public void quoteStringLiteral(StringBuilder buf, @Nullable String charsetName,
+      String val) {
```
Comment (divyanshuS27) [inline core/src/main/java/org/apache/calcite/sql/dialect/PostgresqlSqlDialect.java:117]: what was the need for this chnage?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -1019,6 +1019,21 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
...
 
+  @LibraryOperator(libraries = {POSTGRESQL})
+  public static final SqlFunction AGE =
+        new SqlFunction("AGE", SqlKind.OTHER_FUNCTION,
+          ReturnTypes.ARG0, null,
+          OperandTypes.family(SqlTypeFamily.DATETIME, SqlTypeFamily.DATETIME),
+          SqlFunctionCategory.TIMEDATE);
+
+  @LibraryOperator(libraries = {POSTGRESQL})
+  public static final SqlFunction REGEXP_MATCHES =
```
Comment (divyanshuS27) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1030]: we already have fucntions like REGEXP_INSTR , REGEXP_LIKE, REGEXP_SIMILAR, REGEXP_CONTAINS .... were none of these sufficient for the use case ?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 4]
```diff
@@ -2307,6 +2322,14 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
           OperandTypes.or(OperandTypes.INTEGER_BOOLEAN, OperandTypes.INTEGER),
           SqlFunctionCategory.TIMEDATE);
 
+  @LibraryOperator(libraries = {ORACLE})
```
Comment (divyanshuS27) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2325]: similar as above we already have a fucntion definition called TO_HEX can that not be used?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (Kowsikraja) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2225]: Yeah @divyanshuS27  we verified it.Actually in spark and postgres function syntax is DATE_TRUNC(string, datetime_types) so the previous implementation returns the first argument type so if it is string its returning string type.We are modifying into datetime types.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 3 above)
Comment (Kowsikraja) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1030]: No @divyanshuS27 REGEXP_MATCHES return a set of text (ARRAY).And above functions are different return types.

File: core/src/main/java/org/apache/calcite/sql/dialect/PostgresqlSqlDialect.java (hunk 2 above)
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/sql/dialect/PostgresqlSqlDialect.java:117]: calcite unparses utf16 characters as a unicode decoded string, which is not required in the case of postgres. thats why this method is overriden. i have added a test case called `testPostgresUnicodeString` please check that it shows the difference b/w calcite dialect unparse and postgres unparse of the same string.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 4 above)
Comment (divyanshuS27) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2325]: yes sorry, I meant FROM_HEX fucntion. Both these functions convert hexadeimal String to Binary data right?

**Review by anuragsaxena2k7 — APPROVED** [review]
Approve

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 4 above)
Comment (Kowsikraja) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: yes both functions representing the same functionality.I have removed HEXTORAW function.Please have a look.

**Pull request merged**